### PR TITLE
RHB: 44

### PR DESCRIPTION
**Pull request by vadimsandutsa**

### CODE REVIEW

File: DevEdu.API/DevEdu.Business/ValidationHelpers/TaskValidationHelper.cs [hunk 1]
```diff
@@ -19,5 +24,21 @@ public void CheckTaskExistence(int taskId)
...
+            List<GroupDto> grByU = new List<GroupDto>();
+            foreach (var group in groupsByTask)
+                grByT.Add(group.Group);
+            foreach (var group in groupsByUser)
+                grByU.Add(group.Group);
+
+            var result = grByT.FirstOrDefault(gt => grByU.Any(gu => gu.Id == gt.Id));
+            if (result == default)
+                throw new AuthorizationException(string.Format(ServiceMessages.UserNotRelatedToTaskMessage, userId, taskId));
+        }
```
Comment (Negodyaj) [inline DevEdu.API/DevEdu.Business/ValidationHelpers/TaskValidationHelper.cs:42]: 1. you are not checking the task existence here, so you need to rename this method
2. GetGroupsByTaskId is a method that returns a list of GroupTaskDto (so this method MUST BE renamed), and you've got to call a method that returns a list of GroupDto by your taskId
3. GetGroupsByUserId that you have written returns the list of UserGroupDto, and you've got to call a method that returns a list of GroupDto by your userId
4. that means that UserGroupDto must be removed (and GetGroupsByUserId will have another implementation)
5. that means that you won't need those ugly foreach-loops
6. variables like grByT and grByU may be part of your PoC during the debug process, but can't be a part of commit to the main branch

File: DevEdu.API/DevEdu.Business/Services/TaskService.cs [hunk 2]
```diff
@@ -9,39 +13,59 @@ public class TaskService : ITaskService
...
             _taskRepository = taskRepository;
             _courseRepository = courseRepository;
             _studentAnswerOnTaskRepository = studentAnswerOnTaskRepository;
+            _taskValidationHelper = taskValidationHelper;
         }
 
-        public TaskDto GetTaskById(int id) => _taskRepository.GetTaskById(id);
-
-        public TaskDto GetTaskWithCoursesById(int id)
+        public TaskDto GetTaskById(int id)
```
Comment (Negodyaj) [inline DevEdu.API/DevEdu.Business/Services/TaskService.cs:31]: I had a "fresh look" on this code and now I have a suggestion for you:
this method must be deleted

File: DevEdu.API/DevEdu.Business/Services/TaskService.cs [hunk 3]
```diff
@@ -9,39 +13,59 @@ public class TaskService : ITaskService
...
         {
+            // check if task exists
+            _taskValidationHelper.CheckTaskExistence(id);
             var taskDto = _taskRepository.GetTaskById(id);
-            taskDto.Courses = _courseRepository.GetCoursesToTaskByTaskId(id);
             return taskDto;
         }
 
-        public TaskDto GetTaskWithAnswersById(int id)
+        public TaskDto GetTaskByIdWithValidation(int id, int userId)
```
Comment (Negodyaj) [inline DevEdu.API/DevEdu.Business/Services/TaskService.cs:39]: rename this method to GetTaskById (as you are using the method from 31st row only here)

File: DevEdu.API/DevEdu.Business/Services/TaskService.cs [hunk 4]
```diff
@@ -9,39 +13,59 @@ public class TaskService : ITaskService
...
         {
-            var taskDto = _taskRepository.GetTaskById(id);
-            taskDto.StudentAnswers = _studentAnswerOnTaskRepository.GetStudentAnswersToTaskByTaskId(id);
+            _taskValidationHelper.CheckTaskExistenceWithValidation(id, userId);
```
Comment (Negodyaj) [inline DevEdu.API/DevEdu.Business/Services/TaskService.cs:41]: 1. check if user exists, and then...
2. check if task exists and modify the method for it to return a dto
3. and only after this make a call to CheckTaskExistenceWithValidation
4. return the dto received in step 2

on step 3, if we found that user doesn't have rights to get this task, we will throw an error
and in this case the dto won't be returned from step 4 
so it's ok to have such implementation

File: DevEdu.API/DevEdu.Business/Services/TaskService.cs [hunk 5]
```diff
@@ -9,39 +13,59 @@ public class TaskService : ITaskService
...
             taskDto.StudentAnswers = _studentAnswerOnTaskRepository.GetStudentAnswersToTaskByTaskId(id);
             return taskDto;
         }
 
-        public List<TaskDto> GetTasks() => _taskRepository.GetTasks();
+        public List<TaskDto> GetTasks()
+        {
+            var tasks = _taskRepository.GetTasks();
+            if (tasks == default)
+                throw new EntityNotFoundException(ServiceMessages.NotFounAnyTaskMessage);
```
Comment (Negodyaj) [inline DevEdu.API/DevEdu.Business/Services/TaskService.cs:66]: don't throw an error here
empty list is a valid result

File: DevEdu.API/DevEdu.Business/Services/TaskService.cs [hunk 6]
```diff
@@ -55,11 +79,16 @@ public int AddTask(TaskDto taskDto)
 
         public TaskDto UpdateTask(TaskDto taskDto)
         {
+            _taskValidationHelper.CheckTaskExistence(taskDto.Id);
```
Comment (Negodyaj) [inline DevEdu.API/DevEdu.Business/Services/TaskService.cs:82]: it's a mistake that you are passing id of the task inside the dto
please take a look at any other update method: id is passed as a parameter
so the line "taskDto.Id = taskId;" must be moved from the controller into this method

File: DevEdu.API/DevEdu.Business/Services/TaskService.cs (hunk 3 above)
Comment (vadimsandutsa) [reply DevEdu.API/DevEdu.Business/Services/TaskService.cs:39]: Not only here, that's why i made those two methods separately

File: DevEdu.API/DevEdu.Business/ValidationHelpers/TaskValidationHelper.cs [hunk 7]
```diff
@@ -1,23 +1,39 @@
...
 
-        public TaskValidationHelper(ITaskRepository taskRepository)
+        public TaskValidationHelper(ITaskRepository taskRepository, IUserRepository userRepository)
         {
             _taskRepository = taskRepository;
+            _userRepository = userRepository;
         }
 
-        public void CheckTaskExistence(int taskId)
+        public TaskDto CheckTaskExistence(int taskId)
```
Comment (Negodyaj) [inline DevEdu.API/DevEdu.Business/ValidationHelpers/TaskValidationHelper.cs:21]: GetTaskByIdAndThrowIfNotFound

File: DevEdu.API/DevEdu.Business/Services/TaskService.cs [hunk 8]
```diff
@@ -1,69 +1,92 @@
...
+        {
+            var tasks = _taskRepository.GetTasks();
+            return tasks;
+        }
+
+        public TaskDto AddTask(TaskDto taskDto)
+        {
+            var taskId = _taskRepository.AddTask(taskDto);
+            if (taskDto.Tags == null || taskDto.Tags.Count == 0)
+                return _taskRepository.GetTaskById(taskId);
```
Comment (Negodyaj) [inline DevEdu.API/DevEdu.Business/Services/TaskService.cs:67]: please provide an ability to create a relation to course or group

File: DevEdu.API/DevEdu.Business/Services/TaskService.cs [hunk 9]
```diff
@@ -1,69 +1,92 @@
...
+
+            taskDto.Tags.ForEach(tag => AddTagToTask(taskId, tag.Id));
+            return _taskRepository.GetTaskById(taskId);
+        }
+
+        public TaskDto UpdateTask(TaskDto taskDto, int taskId)
+        {
+            _taskValidationHelper.CheckTaskExistence(taskDto.Id);
+            taskDto.Id = taskId;
+            _taskRepository.UpdateTask(taskDto);
```
Comment (Negodyaj) [inline DevEdu.API/DevEdu.Business/Services/TaskService.cs:77]: please add a validation for userId and taskId relation

File: DevEdu.API/DevEdu.Business/Services/TaskService.cs [hunk 10]
```diff
@@ -1,69 +1,92 @@
...
+            return _taskRepository.GetTaskById(taskDto.Id);
+        }
+
+        public void DeleteTask(int id)
+        {
+            _taskValidationHelper.CheckTaskExistence(id);
```
Comment (Negodyaj) [inline DevEdu.API/DevEdu.Business/Services/TaskService.cs:83]: the same check

File: DevEdu.API/DevEdu.Business/Services/TaskService.cs [hunk 11]
```diff
@@ -1,69 +1,92 @@
...
+            _taskRepository.DeleteTask(id);
+        }
+
+        public int AddTagToTask(int taskId, int tagId) => _taskRepository.AddTagToTask(taskId, tagId);
```
Comment (Negodyaj) [inline DevEdu.API/DevEdu.Business/Services/TaskService.cs:87]: the same check

File: DevEdu.API/DevEdu.Business/Services/TaskService.cs [hunk 12]
```diff
@@ -1,69 +1,92 @@
...
+
+        public void DeleteTagFromTask(int taskId, int tagId) => _taskRepository.DeleteTagFromTask(taskId, tagId);
```
Comment (Negodyaj) [inline DevEdu.API/DevEdu.Business/Services/TaskService.cs:89]: the same check

**Pull request merged**